File: app/buildz/tmp/kapt3/stubs/debug/com/bits/bee/bpmc/data/repository/PossesRepositoryImpl.java
package com.bits.bee.bpmc.data.repository;

import java.lang.System;

/**
 * Created by aldi on 07/06/22.
 */
@kotlin.Metadata(mv = {1, 7, 1}, k = 1, d1 = {"\u0000`\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0010\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0010\b\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0002\u0010 \n\u0002\b\u0002\n\u0002\u0010\t\n\u0002\b\b\n\u0002\u0018\u0002\n\u0002\b\t\u0018\u00002\u00020\u0001B\u0017\b\u0007\u0012\u0006\u0010\u0002\u001a\u00020\u0003\u0012\u0006\u0010\u0004\u001a\u00020\u0005\u00a2\u0006\u0002\u0010\u0006J1\u0010\u0007\u001a\u00020\b2\u0006\u0010\t\u001a\u00020\n2\u0006\u0010\u000b\u001a\u00020\f2\u0006\u0010\r\u001a\u00020\u000e2\u0006\u0010\u000f\u001a\u00020\u0010H\u0096@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\u0011J\u0019\u0010\u0012\u001a\n\u0012\u0006\u0012\u0004\u0018\u00010\u00140\u0013H\u0096@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\u0015J#\u0010\u0016\u001a\u0014\u0012\u0010\u0012\u000e\u0012\n\u0012\b\u0012\u0004\u0012\u00020\u00140\u00180\u00170\u0013H\u0096@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\u0015J*\u0010\u0019\u001a\u0014\u0012\u0010\u0012\u000e\u0012\n\u0012\b\u0012\u0004\u0012\u00020\u00140\u00180\u00170\u00132\u0006\u0010\u001a\u001a\u00020\u001b2\u0006\u0010\u001c\u001a\u00020\u001bH\u0016J*\u0010\u001d\u001a\u0014\u0012\u0010\u0012\u000e\u0012\n\u0012\b\u0012\u0004\u0012\u00020\u00140\u00180\u00170\u00132\u0006\u0010\u001a\u001a\u00020\u001b2\u0006\u0010\u001c\u001a\u00020\u001bH\u0016J*\u0010\u001e\u001a\u0014\u0012\u0010\u0012\u000e\u0012\n\u0012\b\u0012\u0004\u0012\u00020\u00140\u00180\u00170\u00132\u0006\u0010\u001a\u001a\u00020\u001b2\u0006\u0010\u001c\u001a\u00020\u001bH\u0016J*\u0010\u001f\u001a\u0014\u0012\u0010\u0012\u000e\u0012\n\u0012\b\u0012\u0004\u0012\u00020\u00140\u00180\u00170\u00132\u0006\u0010 \u001a\u00020\u001b2\u0006\u0010!\u001a\u00020\u001bH\u0016J\u001a\u0010\"\u001a\u0014\u0012\u0010\u0012\u000e\u0012\n\u0012\b\u0012\u0004\u0012\u00020\u00140\u00180\u00170\u0013H\u0016J\u001d\u0010#\u001a\u000e\u0012\n\u0012\b\u0012\u0004\u0012\u00020\u00140$0\u0013H\u0096@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\u0015J\u001c\u0010%\u001a\u000e\u0012\n\u0012\b\u0012\u0004\u0012\u00020\u00140\u00170\u00132\u0006\u0010&\u001a\u00020\fH\u0016J\u001d\u0010\'\u001a\u000e\u0012\n\u0012\b\u0012\u0004\u0012\u00020\u00140$0\u0013H\u0096@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\u0015J\u001a\u0010(\u001a\u0014\u0012\u0010\u0012\u000e\u0012\n\u0012\b\u0012\u0004\u0012\u00020\u00140\u00180\u00170\u0013H\u0016J\u001a\u0010)\u001a\u0014\u0012\u0010\u0012\u000e\u0012\n\u0012\b\u0012\u0004\u0012\u00020\u00140\u00180\u00170\u0013H\u0016J\u0019\u0010*\u001a\u00020\b2\u0006\u0010+\u001a\u00020\u0014H\u0096@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010,R\u000e\u0010\u0004\u001a\u00020\u0005X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0002\u001a\u00020\u0003X\u0082\u0004\u00a2\u0006\u0002\n\u0000\u0082\u0002\u0004\n\u0002\b\u0019\u00a8\u0006-"}, d2 = {"Lcom/bits/bee/bpmc/data/repository/PossesRepositoryImpl;", "Lcom/bits/bee/bpmc/domain/repository/PossesRepository;", "possesDao", "Lcom/bits/bee/bpmc/data/data_source/local/dao/PossesDao;", "defaultDispatcher", "Lkotlinx/coroutines/CoroutineDispatcher;", "(Lcom/bits/bee/bpmc/data/data_source/local/dao/PossesDao;Lkotlinx/coroutines/CoroutineDispatcher;)V", "addPosses", "", "startBal", "Ljava/math/BigDecimal;", "shift", "", "branch", "Lcom/bits/bee/bpmc/domain/model/Branch;", "cashier", "Lcom/bits/bee/bpmc/domain/model/Cashier;", "(Ljava/math/BigDecimal;ILcom/bits/bee/bpmc/domain/model/Branch;Lcom/bits/bee/bpmc/domain/model/Cashier;Lkotlin/coroutines/Continuation;)Ljava/lang/Object;", "getActivePosses", "Lkotlinx/coroutines/flow/Flow;", "Lcom/bits/bee/bpmc/domain/model/Posses;", "(Lkotlin/coroutines/Continuation;)Ljava/lang/Object;", "getActivePossesList", "Lcom/bits/bee/bpmc/utils/Resource;", "", "getFilter", "current", "", "end", "getFilterAsc", "getFilterDesc", "getJmlPossesByDate", "startDate", "endDate", "getListPossesHistory", "getPosses", "Landroidx/paging/PagingData;", "getPossesById", "id", "getPossesHistory", "getSortAsc", "getSortDesc", "updatePosses", "posses", "(Lcom/bits/bee/bpmc/domain/model/Posses;Lkotlin/coroutines/Continuation;)Ljava/lang/Object;", "app_debug"})
public final class PossesRepositoryImpl implements com.bits.bee.bpmc.domain.repository.PossesRepository {
    private final com.bits.bee.bpmc.data.data_source.local.dao.PossesDao possesDao = null;
    private final kotlinx.coroutines.CoroutineDispatcher defaultDispatcher = null;
    
    @javax.inject.Inject()
    public PossesRepositoryImpl(@org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.data.data_source.local.dao.PossesDao possesDao, @org.jetbrains.annotations.NotNull()
    kotlinx.coroutines.CoroutineDispatcher defaultDispatcher) {
        super();
    }
    
    @org.jetbrains.annotations.Nullable()
    @java.lang.Override()
    public java.lang.Object getActivePosses(@org.jetbrains.annotations.NotNull()
    kotlin.coroutines.Continuation<? super kotlinx.coroutines.flow.Flow<com.bits.bee.bpmc.domain.model.Posses>> continuation) {
        return null;
    }
    
    @org.jetbrains.annotations.Nullable()
    @java.lang.Override()
    public java.lang.Object getPosses(@org.jetbrains.annotations.NotNull()
    kotlin.coroutines.Continuation<? super kotlinx.coroutines.flow.Flow<androidx.paging.PagingData<com.bits.bee.bpmc.domain.model.Posses>>> continuation) {
        return null;
    }
    
    @org.jetbrains.annotations.Nullable()
    @java.lang.Override()
    public java.lang.Object addPosses(@org.jetbrains.annotations.NotNull()
    java.math.BigDecimal startBal, int shift, @org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.domain.model.Branch branch, @org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.domain.model.Cashier cashier, @org.jetbrains.annotations.NotNull()
    kotlin.coroutines.Continuation<? super kotlin.Unit> continuation) {
        return null;
    }
    
    @org.jetbrains.annotations.Nullable()
    @java.lang.Override()
    public java.lang.Object updatePosses(@org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.domain.model.Posses posses, @org.jetbrains.annotations.NotNull()
    kotlin.coroutines.Continuation<? super kotlin.Unit> continuation) {
        return null;
    }
    
    @org.jetbrains.annotations.NotNull()
    @java.lang.Override()
    public kotlinx.coroutines.flow.Flow<com.bits.bee.bpmc.utils.Resource<com.bits.bee.bpmc.domain.model.Posses>> getPossesById(int id) {
        return null;
    }
    
    @org.jetbrains.annotations.Nullable()
    @java.lang.Override()
    public java.lang.Object getPossesHistory(@org.jetbrains.annotations.NotNull()
    kotlin.coroutines.Continuation<? super kotlinx.coroutines.flow.Flow<androidx.paging.PagingData<com.bits.bee.bpmc.domain.model.Posses>>> continuation) {
        return null;
    }
    
    @org.jetbrains.annotations.Nullable()
    @java.lang.Override()
    public java.lang.Object getActivePossesList(@org.jetbrains.annotations.NotNull()
    kotlin.coroutines.Continuation<? super kotlinx.coroutines.flow.Flow<? extends com.bits.bee.bpmc.utils.Resource<? extends java.util.List<com.bits.bee.bpmc.domain.model.Posses>>>> continuation) {
        return null;
    }
    
    @org.jetbrains.annotations.NotNull()
    @java.lang.Override()
    public kotlinx.coroutines.flow.Flow<com.bits.bee.bpmc.utils.Resource<java.util.List<com.bits.bee.bpmc.domain.model.Posses>>> getSortDesc() {
        return null;
    }
    
    @org.jetbrains.annotations.NotNull()
    @java.lang.Override()
    public kotlinx.coroutines.flow.Flow<com.bits.bee.bpmc.utils.Resource<java.util.List<com.bits.bee.bpmc.domain.model.Posses>>> getSortAsc() {
        return null;
    }
    
    @org.jetbrains.annotations.NotNull()
    @java.lang.Override()
    public kotlinx.coroutines.flow.Flow<com.bits.bee.bpmc.utils.Resource<java.util.List<com.bits.bee.bpmc.domain.model.Posses>>> getFilter(long current, long end) {
        return null;
    }
    
    @org.jetbrains.annotations.NotNull()
    @java.lang.Override()
    public kotlinx.coroutines.flow.Flow<com.bits.bee.bpmc.utils.Resource<java.util.List<com.bits.bee.bpmc.domain.model.Posses>>> getListPossesHistory() {
        return null;
    }
    
    @org.jetbrains.annotations.NotNull()
    @java.lang.Override()
    public kotlinx.coroutines.flow.Flow<com.bits.bee.bpmc.utils.Resource<java.util.List<com.bits.bee.bpmc.domain.model.Posses>>> getJmlPossesByDate(long startDate, long endDate) {
        return null;
    }
    
    @org.jetbrains.annotations.NotNull()
    @java.lang.Override()
    public kotlinx.coroutines.flow.Flow<com.bits.bee.bpmc.utils.Resource<java.util.List<com.bits.bee.bpmc.domain.model.Posses>>> getFilterAsc(long current, long end) {
        return null;
    }
    
    @org.jetbrains.annotations.NotNull()
    @java.lang.Override()
    public kotlinx.coroutines.flow.Flow<com.bits.bee.bpmc.utils.Resource<java.util.List<com.bits.bee.bpmc.domain.model.Posses>>> getFilterDesc(long current, long end) {
        return null;
    }
}

File: app/src/main/java/com/bits/bee/bpmc/data/repository/PossesRepositoryImpl.kt
package com.bits.bee.bpmc.data.repository

import androidx.paging.Pager
import androidx.paging.PagingConfig
import androidx.paging.PagingData
import androidx.paging.map
import com.bits.bee.bpmc.data.data_source.local.dao.PossesDao
import com.bits.bee.bpmc.data.data_source.local.model.PossesEntity
import com.bits.bee.bpmc.domain.mapper.PossesDataMapper
import com.bits.bee.bpmc.domain.model.Branch
import com.bits.bee.bpmc.domain.model.Cashier
import com.bits.bee.bpmc.domain.model.Posses
import com.bits.bee.bpmc.domain.model.User
import com.bits.bee.bpmc.domain.repository.PossesRepository
import com.bits.bee.bpmc.utils.BPMConstants
import com.bits.bee.bpmc.utils.Resource
import com.bits.bee.bpmc.utils.TrxNoGeneratorUtils
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.flow.mapLatest
import kotlinx.coroutines.withContext
import java.math.BigDecimal
import java.util.*
import javax.inject.Inject

/**
 * Created by aldi on 07/06/22.
 */
class PossesRepositoryImpl @Inject constructor(
    private val possesDao: PossesDao,
    private val defaultDispatcher: CoroutineDispatcher
) : PossesRepository {

    override suspend fun getActivePosses(): Flow<Posses?> = flow {
        val posses = possesDao.getActivePosses()
        emit(posses?.let{PossesDataMapper.fromDbToDomain(posses)})
    }.flowOn(defaultDispatcher)

    override suspend fun getPosses(): Flow<PagingData<Posses>> = Pager(
        config = PagingConfig(
            pageSize = BPMConstants.BPM_LIMIT_PAGINATION,
            maxSize = BPMConstants.BPM_MAX_PAGINATION,
            enablePlaceholders = true
        ),
        pagingSourceFactory = {
            possesDao.getListPosses()
        }
    ).flow.mapLatest {
        it.map { PossesDataMapper.fromDbToDomain(it) }
    }.flowOn(defaultDispatcher)

    override suspend fun getLastPosses(): Flow<Posses?> = flow {
        val posses = possesDao.getLastPosses()
        emit(posses?.let{PossesDataMapper.fromDbToDomain(posses)})
    }.flowOn(defaultDispatcher)

    override suspend fun addPosses(startBal: BigDecimal, shift : Int, branch: Branch, cashier: Cashier, user : User) {
        val date = Date()
        withContext(defaultDispatcher) {
            val posses = PossesEntity(
                startBal = startBal,
                trxDate = date,
                cashierId = cashier.id,
                trxNo = TrxNoGeneratorUtils.generatePossesTrxNo(shift, cashier),
                startTime = date,
                shift = shift,
                total = BigDecimal.ZERO,
                endBal = null,
                endCash = BigDecimal.ZERO,
                credit = BigDecimal.ZERO,
                userId = user.id,
                totalActualCash = BigDecimal.ZERO,
                totalDiffCash = BigDecimal.ZERO,
                isUploaded = false
            )
            possesDao.insertSingle(posses)
        }
    }

    override suspend fun updatePosses(posses: Posses) {
        withContext(defaultDispatcher){
            possesDao.update(PossesDataMapper.fromDomainToDb(posses))
        }
    }

    override fun getPossesById(id: Int): Flow<Resource<Posses>> {
        return flow {
            val data = possesDao.getPossesById(id)
            emit(Resource.success(PossesDataMapper.fromDbToDomain(data)))
        }.flowOn(defaultDispatcher)
    }

    override suspend fun getPossesHistory(): Flow<PagingData<Posses>> = Pager(
        config = PagingConfig(
            pageSize = BPMConstants.BPM_LIMIT_PAGINATION,
            maxSize = BPMConstants.BPM_MAX_PAGINATION,
            enablePlaceholders = true
        ),
        pagingSourceFactory = {
            possesDao.getPossesHistory()
        }
    ).flow.mapLatest {
        it.map { PossesDataMapper.fromDbToDomain(it) }
    }.flowOn(defaultDispatcher)

    override fun getFilter(current: Long, end: Long): Flow<Resource<List<Posses>>> {
        return flow {
            val data = possesDao.getFilter(current, end).map { PossesDataMapper.fromDbToDomain(it) }
            emit(Resource.success(data))
        }.flowOn(defaultDispatcher)
    }

    override fun getListPossesHistory(): Flow<Resource<List<Posses>>> {
        return flow {
            val data = possesDao.getListPossesHistory().map { PossesDataMapper.fromDbToDomain(it) }
            emit(Resource.success(data))
        }.flowOn(defaultDispatcher)
    }

    override fun getJmlPossesByDate(startDate: Long, endDate: Long): Flow<List<Posses>> {
        return flow {
           val data = possesDao.getJmlPossesByDate(startDate, endDate).map { PossesDataMapper.fromDbToDomain(it) }
            emit(data)
        }.flowOn(defaultDispatcher)
    }

    override fun getFilterAsc(current: Long, end: Long): Flow<Resource<List<Posses>>> {
        return flow {
            val data = possesDao.getFilterAsc(current, end).map { PossesDataMapper.fromDbToDomain(it) }
            emit(Resource.success(data))
        }.flowOn(defaultDispatcher)
    }

    override fun getFilterDesc(current: Long, end: Long): Flow<Resource<List<Posses>>> {
        return flow {
            val data = possesDao.getFilterDesc(current, end).map { PossesDataMapper.fromDbToDomain(it) }
            emit(Resource.success(data))
        }.flowOn(defaultDispatcher)
    }

    override fun getPossesHaventUpload(): Flow<List<Posses>> {
        return flow {
            val data = possesDao.getPossesHaventUpload().map { PossesDataMapper.fromDbToDomain(it) }
            emit((data))
        }.flowOn(defaultDispatcher)
    }

}